README.md
[![CircleCI](https://circleci.com/gh/treasure-data/embulk-input-google_analytics/tree/master.svg?style=svg)](https://circleci.com/gh/treasure-data/embulk-input-google_analytics/tree/master)
[![Code Climate](https://codeclimate.com/github/treasure-data/embulk-input-google_analytics/badges/gpa.svg)](https://codeclimate.com/github/treasure-data/embulk-input-google_analytics)
[![Test Coverage](https://codeclimate.com/github/treasure-data/embulk-input-google_analytics/badges/coverage.svg)](https://codeclimate.com/github/treasure-data/embulk-input-google_analytics/coverage)
[![Issue Count](https://codeclimate.com/github/treasure-data/embulk-input-google_analytics/badges/issue_count.svg)](https://codeclimate.com/github/treasure-data/embulk-input-google_analytics)
[![Gem Version](https://badge.fury.io/rb/embulk-input-google_analytics.svg)](https://badge.fury.io/rb/embulk-input-google_analytics)

# Google Analytics input plugin for Embulk

Embulk input plugin for Google Analytics reports.

## Configuration

- **json_key_content**: See example config.
- **view_id**: View ID for target data. See [Get View ID](#get-view-id) (string, required)
- **time_series**: Only `ga:dateHour` or `ga:date` (string, required)
- **dimensions**: Target dimensions (array, default: `[]` )
- **metrics**: Target metrics (array, default: `[]` )
- **start_date**: Target report start date. Valid format is "YYYY-MM-DD". (string, default: [7 days ago](https://developers.google.com/analytics/devguides/reporting/core/v4/rest/v4/reports/batchGet#reportrequest))
- **end_date**: Target report end date. Valid format is "YYYY-MM-DD". (string, default: [1 day ago](https://developers.google.com/analytics/devguides/reporting/core/v4/rest/v4/reports/batchGet#reportrequest))
- **incremental**: `true` for generate "config_diff" with `embulk run -c config.diff` (bool, default: true)
- **last_record_time**: Ignore fetched records until this time. Mainly for incremental:true. (string, default: nil)
- **retry_limit**: Try to retry this times (integer, default: 5)
- **retry_initial_wait_sec**: Wait seconds for exponential backoff initial value (integer, default: 2)

### **New update from verions  0.1.18**
Started from version 0.1.18, the Plugin also supports User Account Authentication along with Service Account Authentication see: [OAuth 2.0 for Server-side Web Application](https://developers.google.com/identity/protocols/OAuth2WebServer). Extra optional configuration keys were added and the **json_key_content** is made optional
 - **client_id**: client_id for application (string, optional)
 - **client_secret**: client_secret for application (string, optional)
 - **refresh_token**: the refresh_token obtained during exchange authentication code (string, optional)

### Get View ID

1. Go to the [Google Analytics sign in page](https://analytics.google.com/analytics/) and sign in.
1. Click "Admin" tab at left below
1. Select the "Property" using the drop-down menu below ‘Property’.
1. Select ‘View Settings’ beneath ‘View’.
1. The View ID for the selected property is listed first under ‘Basic Settings

### About `json_key_content` option.

You need a service account on Google.

<ol>
  <li>Open the <a href="https://console.developers.google.com/permissions/serviceaccounts"><b>Service accounts</b> page</a>. If prompted,
select a project.</li>
  <li>Click <b>Create service account</b>.</li>
  <li>

    In the <b>Create service account</b> window, type a name for the service
    account, and select <b>Furnish a new private key</b>. If you want to
    <a href="https://developers.google.com/identity/protocols/OAuth2ServiceAccount#delegatingauthority">grant
    Google Apps domain-wide authority</a> to the service account, also select
    <b>Enable Google Apps Domain-wide Delegation</b>.

    Then click <b>Create</b>.</li>
</ol>
From: <https://developers.google.com/identity/protocols/OAuth2ServiceAccount>

Screenshot: ![Service Account](./service_account.png)

## Why the result doesn't match with web interface?

Google Reporting API uses "sampling" data.

- https://developers.google.com/analytics/devguides/reporting/core/v4/basics#sampling
- https://support.google.com/analytics/answer/2637192

That means sometimes result will be unmatched with Google Analytics web interface, and the result is based on sampled data, not all of raw data. This is a Google API's limitation.

Currently a sampling level supported by this plugin is DEFAULT only. Let us know if you want to use other sampling level (SMALL or LARGE).

## Example

```yaml
in:
  type: google_analytics
  json_key_content: |
    {
      "type": "service_account",
      "project_id": "....",
      "private_key_id": "....",
      "private_key": "-----BEGIN PRIVATE KEY-----\n..........................\n-----END PRIVATE KEY-----\n",
      "client_email": ".....",
      "client_id": ".........",
      "auth_uri": "https://accounts.google.com/o/oauth2/auth",
      "token_uri": "https://accounts.google.com/o/oauth2/token",
      "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
      "client_x509_cert_url": ".........."
    }
  view_id: 123111111
  time_series: "ga:dateHour" # hourly basis

  # https://developers.google.com/analytics/devguides/reporting/core/dimsmets
  dimensions:
    - "ga:browser"
  metrics:
    - "ga:visits"
    - "ga:pageviews"

  start_date: "2016-06-27"
  end_date: "2016-06-28"
```

## Config example using User Authentication
```yaml
 in:
  type: google_analytics
  client_id: "#############apps.googleusercontent.com"
  client_secret: "##############QLxgrfis4"
  refresh_token: "##########awWNT9lTeGq8weKE"
  view_id: 123111111
  time_series: "ga:dateHour" # hourly basis
  dimensions:
    - "ga:browser"
  metrics:
    - "ga:visits"
    - "ga:pageviews"

  start_date: "2016-06-27"
  end_date: "2016-06-28"
```

## Build

```
$ rake build
```


CHANGELOG.md
## 0.1.24 - 2020-07-14
Upgrade version of gems to be compatible with other google input plugins [#46](https://github.com/treasure-data/embulk-input-google_analytics/pull/46)
Release version 0.1.24 https://github.com/treasure-data/embulk-input-google_analytics/commit/3aef789a672227ae3ad9b58235091d34c96a4b67

## 0.1.23 - 2019-10-24
Lock signet and activesupport by less than or equal locked version [#45](https://github.com/treasure-data/embulk-input-google_analytics/pull/45)

## 0.1.22 - 2019-10-24
* Fix **Plugin's dependencies require Ruby > 2.3** [#44](https://github.com/treasure-data/embulk-input-google_analytics/pull/44)

## 0.1.21 - 2019-04-01
* Fix **(ArgumentError) no time information in ""** when last_record_time is null [#42](https://github.com/treasure-data/embulk-input-google_analytics/pull/42)

## 0.1.19 - 2018-10-03
* Make it easy to understand error message of Google Analytics connector

## 0.1.18 - 2018-05-23
* Suport User Account Authentication [#33](https://github.com/treasure-data/embulk-input-google_analytics/pull/33)
* Ignore last_record_time if non-incremental task [#34](https://github.com/treasure-data/embulk-input-google_analytics/pull/34)

## 0.1.17 - 2018-04-16

* Add more descriptive error message when Google API return invalid time format [#32](https://github.com/treasure-data/embulk-input-google_analytics/pull/32)

## 0.1.16 - 2018-02-06

* Use user profile timezone when format next from_date [#31](https://github.com/treasure-data/embulk-input-google_analytics/pull/31)


## 0.1.15 - 2017-11-29

* Add explicit logging for the number of skipped records [#30](https://github.com/treasure-data/embulk-input-google_analytics/pull/30)

## 0.1.14 - 2017-06-27

* Fixed version of google-api-client [#26](https://github.com/treasure-data/embulk-input-google_analytics/pull/26)

## 0.1.13 - 2017-03-28
* Fixed bug `last_record_time` in `preview` mode [#20](https://github.com/treasure-data/embulk-input-google_analytics/pull/20)

## 0.1.12 - 2017-01-31
* Support 'fooXXbar' column names [#19](https://github.com/treasure-data/embulk-input-google_analytics/pull/19)

## 0.1.11 - 2017-01-16
* Throw DataError if '(other)' values appear as ga:dateHour as ga:date or ga:dateHour [#18](https://github.com/treasure-data/embulk-input-google_analytics/pull/18)

## 0.1.10 - 2016-12-08
* Logging unexpected time string [#17](https://github.com/treasure-data/embulk-input-google_analytics/pull/17)

## 0.1.9 - 2016-11-22
* Support 'foobarXX' column names [#16](https://github.com/treasure-data/embulk-input-google_analytics/pull/16)

## 0.1.8 - 2016-11-09
* TIME column should be double, not timestamp. e.g. sessionDuration (thx @kazuya030) [#13](https://github.com/treasure-data/embulk-input-google_analytics/pull/13) [#15](https://github.com/treasure-data/embulk-input-google_analytics/pull/15)

## 0.1.7 - 2016-10-20
* Fix to generate `end_date` on config_diff  [#12](https://github.com/treasure-data/embulk-input-google_analytics/pull/12)

## 0.1.6 - 2016-08-29
* Add no-op guessing to avoid guess error [#11](https://github.com/treasure-data/embulk-input-google_analytics/pull/11)

## 0.1.5 - 2016-08-23
* Enable custom dimensions [#10](https://github.com/treasure-data/embulk-input-google_analytics/pull/10)

## 0.1.4 - 2016-08-19
* Always add `view_id` into records [#9](https://github.com/treasure-data/embulk-input-google_analytics/pull/9)

## 0.1.3 - 2016-08-15
* Use net/http to avoid TLS error [#8](https://github.com/treasure-data/embulk-input-google_analytics/pull/8)

## 0.1.2 - 2016-07-13

* Fix when `end_date` is nil [#7](https://github.com/treasure-data/embulk-input-google_analytics/pull/7)

## 0.1.1 - 2016-07-13
* Enable scheduled execution [#4](https://github.com/treasure-data/embulk-input-google_analytics/pull/4)
* Error handling [#6](https://github.com/treasure-data/embulk-input-google_analytics/pull/6)
* Ignore too early accessing data due to it is not fixed value [#5](https://github.com/treasure-data/embulk-input-google_analytics/pull/5)

## 0.1.0 - 2016-07-07

The first release!!


test/run-test.rb
#!/usr/bin/env ruby

base_dir = File.expand_path(File.join(File.dirname(__FILE__), ".."))
lib_dir = File.join(base_dir, "lib")
test_dir = File.join(base_dir, "test")

require "test-unit"
require "test/unit/rr"

$LOAD_PATH.unshift(lib_dir)
$LOAD_PATH.unshift(test_dir)

ENV["TEST_UNIT_MAX_DIFF_TARGET_STRING_SIZE"] ||= "5000"

if ENV["COVERAGE"]
  require 'simplecov'
  SimpleCov.start 'test_frameworks'
end

exit Test::Unit::AutoRunner.run(true, test_dir)


test/fixture_helper.rb
require "yaml"
require "pathname"

module FixtureHelper
  def fixture_dir
    Pathname.new(__FILE__).dirname.join("fixtures")
  end

  def fixture_read(name)
    fixture_dir.join(name).read
  end

  def fixture_load(name)
    YAML.load fixture_read(name)
  end
end


test/override_assert_raise.rb
module OverrideAssertRaise
  # NOTE: Embulk 0.7.1+ required to raise ConfigError to do as `ConfigError.new("message")`,
  #       original `assert_raise` method can't catch that, but `begin .. rescue` can.
  #       So we override assert_raise as below.
  def assert_raise(expected_class_or_instance = StandardError, &block)
    begin
      expected_class = expected_class_or_instance.is_a?(Class) ? expected_class_or_instance : expected_class_or_instance.class
      block.call
      assert_equal expected_class, nil
    rescue ::Test::Unit::AssertionFailedError => e
      # failed assert raises this Error and that extends StandardError, so rescue it first
      raise e
    rescue expected_class => e
      # https://github.com/test-unit/test-unit/issues/94
      if expected_class_or_instance == expected_class
        assert true # passed
      else
        assert_equal expected_class_or_instance.message, e.message
      end
    rescue => e
      assert_equal(expected_class, e.class) # not expected one raised
    end
  end
end


lib/embulk/input/google_analytics/client.rb
require "perfect_retry"
require "active_support/core_ext/time"
require "google/apis/analyticsreporting_v4"
require "google/apis/analytics_v3"

module Embulk
  module Input
    module GoogleAnalytics
      class Client
        attr_reader :task

        def initialize(task, is_preview = false)
          @task = task
          @is_preview = is_preview
        end

        def preview?
          @is_preview
        end

        def each_report_row(&block)
          page_token = nil
          Embulk.logger.info "view_id:#{view_id} timezone has been set as '#{get_profile[:timezone]}'"

          loop do
            result = get_reports(page_token)
            report = result.to_h[:reports].first

            if !report[:data].has_key?(:rows)
              Embulk.logger.warn "Result doesn't contain rows."
              break
            end

            if report[:data][:rows].empty?
              Embulk.logger.warn "Result has 0 rows."
              break
            end

            dimensions = report[:column_header][:dimensions]
            metrics = report[:column_header][:metric_header][:metric_header_entries].map{|m| m[:name]}
            report[:data][:rows].each do |row|
              dim = dimensions.zip(row[:dimensions]).to_h
              met = metrics.zip(row[:metrics].first[:values]).to_h
              format_row = dim.merge(met)
              raw_time = format_row[task["time_series"]]
              optimize_value_by_query_limit?(raw_time)
              next if too_early_data?(raw_time)
              format_row[task["time_series"]] = time_parse_with_profile_timezone(raw_time)
              format_row["view_id"] = view_id
              block.call format_row
            end

            break if preview?

            unless page_token = report[:next_page_token]
              break
            end
            Embulk.logger.info "Fetching report with page_token: #{page_token}"
          end
        end

        def get_profile
          @profile ||=
            begin
              profile = get_all_profiles.to_h[:items].find do |prof|
                prof[:id] == view_id
              end

              unless profile
                raise Embulk::ConfigError.new("Can't find view_id:#{view_id} profile via Google Analytics API.")
              end

              profile
            end
        end

        def get_all_profiles
          service = Google::Apis::AnalyticsV3::AnalyticsService.new  # TODO: weida check here
          service.authorization = auth

          Embulk.logger.debug "Fetching profile from API"
          retryer.with_retry do
            service.list_profiles("~all", "~all")
          end
        end

        def optimize_value_by_query_limit?(data)
          # For any date range, Analytics returns a maximum of 1 million rows for the report. Rows in excess of 1 million are rolled-up into an (other) row.
          # See more details: https://support.google.com/analytics/answer/1009671
          if data.to_s == "(other)"
            raise Embulk::DataError.new('Stop fetching data from Analytics because over 1M data fetching was limited. Please reduce data range to fetch data according to this article: https://support.google.com/analytics/answer/1009671.')
          end
        end

        def time_parse_with_profile_timezone(time_string)
          date_format =
            case task["time_series"]
            when "ga:dateHour"
              "%Y%m%d%H"
            when "ga:date"
              "%Y%m%d"
            end
          parts = Date._strptime(time_string, date_format)
          unless parts
            # strptime was failed. Google API returns unexpected date string.
            raise Embulk::DataError.new("Failed to parse #{task["time_series"]} data. The value is '#{time_string}'(#{time_string.class}) and it doesn't match with '#{date_format}'.")
          end

          swap_time_zone do
            Time.zone.local(*parts.values_at(:year, :mon, :mday, :hour)).to_time
          end
        end

        def get_reports(page_token = nil)
          # https://developers.google.com/analytics/devguides/reporting/core/v4/rest/v4/reports/batchGet
          service = Google::Apis::AnalyticsreportingV4::AnalyticsReportingService.new
          service.authorization = auth

          request = Google::Apis::AnalyticsreportingV4::GetReportsRequest.new
          request.report_requests = build_report_request(page_token)

          Embulk.logger.info "Query to Core Report API: #{request.to_json}"
          retryer.with_retry do
            service.batch_get_reports request
          end
        end

        def get_columns_list
          columns = get_custom_dimensions + get_metadata_columns
          canonical_column_names(columns)
        end

        def canonical_column_names(columns)
          result = []
          columns.each do |col|
            if col[:id].match(/XX/)
              # for such columns:
              # https://developers.google.com/analytics/devguides/reporting/core/dimsmets#view=detail&group=content_grouping
              # https://developers.google.com/analytics/devguides/reporting/metadata/v3/devguide#attributes
              min = [
                col[:attributes][:minTemplateIndex],
                col[:attributes][:premiumMinTemplateIndex],
              ].compact.min
              max = [
                col[:attributes][:maxTemplateIndex],
                col[:attributes][:premiumMaxTemplateIndex],
              ].compact.max

              min.upto(max) do |n|
                actual_id = col[:id].gsub(/XX/, n.to_s)
                result << col.merge(id: actual_id)
              end
            else
              result << col
            end
          end
          result
        end

        def get_metadata_columns
          # https://developers.google.com/analytics/devguides/reporting/metadata/v3/reference/metadata/columns/list
          service = Google::Apis::AnalyticsV3::AnalyticsService.new
          service.authorization = auth
          retryer.with_retry do
            service.list_metadata_columns("ga").to_h[:items]
          end
        end

        def get_custom_dimensions
          # https://developers.google.com/analytics/devguides/config/mgmt/v3/mgmtReference/management/customDimensions/list
          service = Google::Apis::AnalyticsV3::AnalyticsService.new
          service.authorization = auth
          retryer.with_retry do
            service.list_custom_dimensions(get_profile[:account_id], get_profile[:web_property_id]).to_h[:items]
          end
        end

        def build_report_request(page_token = nil)
          query = {
            view_id: view_id,
            dimensions: [{name: task["time_series"]}] + task["dimensions"].map{|d| {name: d}},
            metrics: task["metrics"].map{|m| {expression: m}},
            include_empty_rows: true,
            page_size: preview? ? 10 : 10000,
          }

          if task["start_date"] || task["end_date"]
            query[:date_ranges] = [{
              start_date: task["start_date"],
              end_date: task["end_date"],
            }]
          end

          if page_token
            query[:page_token] = page_token
          end

          [query]
        end

        def view_id
          task["view_id"]
        end

        def auth
          retryer.with_retry do
            case task['auth_method']
            when Plugin::AUTH_TYPE_JSON_KEY
              Google::Auth::ServiceAccountCredentials.make_creds(
                json_key_io: StringIO.new(task["json_key_content"]),
                scope: "https://www.googleapis.com/auth/analytics.readonly"
              )
            when Plugin::AUTH_TYPE_REFRESH_TOKEN
              Google::Auth::UserRefreshCredentials.new(
                'token_credential_uri': Google::Auth::UserRefreshCredentials::TOKEN_CRED_URI,
                'client_id': task['client_id'],
                'client_secret': task['client_secret'],
                'refresh_token': task['refresh_token']
              )
            else
              raise Embulk::ConfigError.new("Unknown Authentication method: '#{task['auth_method']}'.")
            end
          end
        rescue Google::Apis::AuthorizationError => e
          raise ConfigError.new(e.message)
        end

        def swap_time_zone(&block)
          orig_timezone = Time.zone
          Time.zone = get_profile[:timezone]
          yield
        ensure
          Time.zone = orig_timezone
        end

        def too_early_data?(time_str)
          # fetching 20160720 data on 2016-07-20, it is too early fetching
          swap_time_zone do
            now = Time.zone.now
            case task["time_series"]
            when "ga:dateHour"
              time_str.to_i >= now.strftime("%Y%m%d%H").to_i
            when "ga:date"
              time_str.to_i >= now.strftime("%Y%m%d").to_i
            end
          end
        end

        def retryer
          PerfectRetry.new do |config|
            config.limit = task["retry_limit"]
            config.logger = Embulk.logger
            config.log_level = nil

            # https://developers.google.com/analytics/devguides/reporting/core/v4/errors
            # https://developers.google.com/analytics/devguides/reporting/core/v4/limits-quotas#additional_quota
            # https://github.com/google/google-api-ruby-client/blob/master/lib/google/apis/errors.rb
            # https://github.com/google/google-api-ruby-client/blob/0.9.11/lib/google/apis/core/http_command.rb#L33
            config.rescues = Google::Apis::Core::HttpCommand::RETRIABLE_ERRORS # TODO: weida check here
            config.dont_rescues = [Embulk::DataError, Embulk::ConfigError]
            config.sleep = lambda{|n| task["retry_initial_wait_sec"]* (2 ** (n-1)) }
            config.raise_original_error = true
          end
        end
      end
    end
  end
end


lib/embulk/input/google_analytics/plugin.rb
require 'uri'

module Embulk
  module Input
    module GoogleAnalytics
      class Plugin < InputPlugin
        ::Embulk::Plugin.register_input("google_analytics", self)
        AUTH_TYPE_JSON_KEY = 'json_key'.freeze
        AUTH_TYPE_REFRESH_TOKEN = 'refresh_token'.freeze

        # https://developers.google.com/analytics/devguides/reporting/core/dimsmets

        def self.transaction(config, &control)
          task = task_from_config(config)
          unless %w(ga:date ga:dateHour ga:yearMonth ga:year).include?(task["time_series"])
            raise ConfigError.new("Unknown time_series '#{task["time_series"]}'. Use 'ga:dateHour', 'ga:date', 'ga:year' or 'ga:yearMonth'")
          end

          raise ConfigError.new("Unknown Authentication method '#{task['auth_method']}'.") unless task['auth_method']

          if task['auth_method'] == Plugin::AUTH_TYPE_REFRESH_TOKEN
            unless task['client_id'] && task['client_secret'] && task['refresh_token']
              raise ConfigError.new("client_id, client_secret and refresh_token are required when using Oauth authentication")
            end
          elsif task['auth_method'] == Plugin::AUTH_TYPE_JSON_KEY
            if !valid_json?(task["json_key_content"])
              raise ConfigError.new("json_key_content is not a valid JSON object")
            end
          end

          columns_list = Client.new(task).get_columns_list

          columns = columns_from_task(task).map do |col_name|
            col_info = columns_list.find{|col| col[:id] == col_name} || {}
            # raise ConfigError.new("Unknown metric/dimension '#{col_name}'") unless col_info

            col_type =
              if col_info[:attributes]
                # standard dimension
                case col_info[:attributes][:dataType]
                when "STRING"
                  :string
                when "INTEGER"
                  :long
                when "PERCENT", "FLOAT", "CURRENCY", "TIME"
                  :double
                end
              else
                # custom dimension
                :string
              end

            # time_series column should be timestamp
            if col_name == task["time_series"]
              col_type = :timestamp
            end
            Column.new(nil, canonicalize_column_name(col_name), col_type)
          end

          columns << Column.new(nil, "view_id", :string)

          resume(task, columns, 1, &control)
        end

        def self.resume(task, columns, count, &control)
          task_reports = yield(task, columns, count)

          next_config_diff = task_reports.first
          return next_config_diff
        end

        def self.task_from_config(config)
          refresh_token = config.param('refresh_token', :string, default: nil)
          json_key_content = config.param("json_key_content", :string, default: nil)

          auth_method = Plugin::AUTH_TYPE_REFRESH_TOKEN if refresh_token
          auth_method = Plugin::AUTH_TYPE_JSON_KEY if json_key_content && auth_method == nil
            {
            "auth_method" => auth_method,
            "client_id" => config.param("client_id", :string, default: nil),
            "client_secret" => config.param("client_secret", :string, default: nil),
            "refresh_token" => refresh_token,
            "json_key_content" => json_key_content,
            "view_id" => config.param("view_id", :string),
            "dimensions" => config.param("dimensions", :array, default: []),
            "metrics" => config.param("metrics", :array, default: []),
            "metric_filters" => config.param("metric_filters", :array, default: []),
            "dimension_filters" => config.param("dimension_filters", :array, default: []),
            "segments" => config.param("segments", :array, default: []),
            "filters_expression" => config.param("filters_expression", :string, default: nil),
            "time_series" => config.param("time_series", :string),
            "start_date" => config.param("start_date", :string, default: nil),
            "end_date" => config.param("end_date", :string, default: nil),
            "incremental" => config.param("incremental", :bool, default: true),
            "last_record_time" => config.param("last_record_time", :string, default: nil),
            "retry_limit" => config.param("retry_limit", :integer, default: 5),
            "retry_initial_wait_sec" => config.param("retry_initial_wait_sec", :integer, default: 2),
            "sampling" => config.param("sampling", :string, default: "DEFAULT"),
          }
        end

        def self.columns_from_task(task)
          [
            task["time_series"],
            task["dimensions"],
            task["metrics"],
          ].flatten.uniq
        end

        def self.canonicalize_column_name(name)
          # ga:dateHour -> date_hour
          name.gsub(/^ga:/, "").gsub(/[A-Z]+/, "_\\0").gsub(/^_/, "").downcase
        end

        def self.guess(config)
          Embulk.logger.warn "Don't needed to guess for this plugin"
          return {}
        end

        def self.valid_json?(json_object)
          # 'null' string is a valid string for parse function
          # However in our case, json_content_key could not be 'null' therefore this check is added
          if json_object == "null"
              return false
          end
          begin
            JSON.parse(json_object)
                return true
          rescue JSON::ParserError => e
              return false
          end
        end

        def init
          if task["start_date"] && !task["end_date"]
            task["end_date"] = "today"
          end
        end

        def run
          client = Client.new(task, preview?)
          columns = self.class.columns_from_task(task) + ["view_id"]

          last_record_time = Time.parse(task["last_record_time"]) if task['incremental'] && !task["last_record_time"].blank?
          latest_time_series = nil
          skip_counter, total_counter = 0, 0
          client.each_report_row do |row|
            time = row[task["time_series"]]
            total_counter += 1
            if !preview? && last_record_time && time <= last_record_time
              skip_counter += 1
              next
            end

            values = row.values_at(*columns)
            page_builder.add values

            latest_time_series = [
              latest_time_series,
              time,
            ].compact.max
          end
          page_builder.finish

          Embulk.logger.info "Total: #{total_counter} rows."
          if skip_counter > 0
            Embulk.logger.info "#{skip_counter} rows were ignored because the rows' date is " +
                                   "before \"last_record_time\": #{last_record_time}."
          end

          if task["incremental"]
            calculate_next_times(client.get_profile[:timezone], latest_time_series)
          else
            {}
          end
        end

        def preview?
          org.embulk.spi.Exec.isPreview()
        rescue java.lang.NullPointerException
          false
        end

        def calculate_next_times(client_time_zone, raw_fetched_latest_time)
          fetched_latest_time = case task["time_series"]
                                when "ga:year", "ga:yearMonth"
                                  parts = [*raw_fetched_latest_time.split('-'), 1, 1].first(3)
                                  Time.new(*parts).to_time
                                else
                                  raw_fetched_latest_time
                                end
          task_report = {}
          if fetched_latest_time
            # Convert fetched_last_time to user timezone
            timezone = ActiveSupport::TimeZone[client_time_zone]
            task_report[:start_date] = timezone.nil? ? fetched_latest_time.strftime("%Y-%m-%d") : timezone.parse(fetched_latest_time.to_s).strftime("%Y-%m-%d")

            # if end_date specified as statically YYYY-MM-DD, it will be conflict with start_date (end_date < start_date)
            # Or when end_date is nil, only start_date will be filled on next run but it is illegal API request.
            # Modify end_date as "today" to be safe
            if task["end_date"].nil? || task["end_date"].match(/[0-9]{4}-[0-9]{2}-[0-9]{2}/)
              task_report[:end_date] = "today" # "today" means now. running at 03:30 AM, will got 3 o'clock data.
            else
              task_report[:end_date] = task["end_date"]
            end

            # "start_date" format is YYYY-MM-DD, but ga:dateHour will return records by hourly.
            # If run at 2016-07-03 05:00:00, start_date will set "2016-07-03" and got records until 2016-07-03 05:00:00.
            # Then next run at 2016-07-04 05:00, will got records between 2016-07-03 00:00:00 and 2016-07-04 05:00:00.
            # It will evantually duplicated between 2016-07-03 00:00:00 and 2016-07-03 05:00:00
            #
            #           Date|        2016-07-03      |   2016-07-04
            #           Hour|    5                   |    5
            # 1st run ------|----|                   |
            # 2nd run       |------------------------|-----
            #               ^^^^^ duplicated
            #
            # "last_record_time" option solves that problem
            #
            #           Date|        2016-07-03      |   2016-07-04
            #           Hour|    5                   |    5
            # 1st run ------|----|                   |
            # 2nd run       #####|-------------------|-----
            #               ^^^^^ ignored (skipped)
            #
            task_report[:last_record_time] = fetched_latest_time.strftime("%Y-%m-%d %H:%M:%S %z")
          else
            # no records fetched, don't modify config_diff
            task_report = {
              start_date: task["start_date"],
              end_date: task["end_date"]
            }
            # write last_record_time only when last_record_time is not nil and not empty
            unless task["last_record_time"].blank?
              task_report[:last_record_time] = task["last_record_time"]
            end 
          end
          task_report
        end
      end
    end
  end
end
